File: 仿写新浪微博/Classes/Tools/PictureBrowser/PictureBrowserViewController.m
//
//  PictureBrowserViewController.m
//  仿写新浪微博OC
//
//  Created by Apple on 2017/8/25.
//  Copyright © 2017年 WRQ. All rights reserved.
//

#import "PictureBrowserViewController.h"
#import "PictureBrowserViewCell.h"
#import "PictureBrowserAnimator.h"
#import "PictureBrowserPhotos.h"
#import "SVProgressHUD.h"

// 可重用cellId
static NSString *const PictureBrowserViewCellId = @"PictureBrowserViewCellId";

@interface PictureBrowserViewController () <UICollectionViewDataSource,PictureBrowserCellDelegate,UICollectionViewDelegate> {
    PictureBrowserPhotos *_photos;
    PictureBrowserAnimator *_animator;
}

/** 页数显示按钮 */
@property(nonatomic, strong) UIButton *pageConutButton;

@end

@implementation PictureBrowserViewController

#pragma mark - 监听方法
- (void)close {
    PictureBrowserViewCell *cell = [_collectionView visibleCells][0];
    _animator.fromImageView = cell.imageView;
    _photos.selectedIndex = [_collectionView indexPathForCell:cell].row;
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)savePicture {
    PictureBrowserViewCell *cell = [_collectionView visibleCells][0];
    
    UIImageWriteToSavedPhotosAlbum(cell.imageView.image, self, @selector(image:didFinishSavingWithError:contextInfo:), nil);
}

- (void)image:(UIImage *)image didFinishSavingWithError:(NSError *)error contextInfo:(void *)contextInfo {
    if (error == nil) {
        [SVProgressHUD showSuccessWithStatus:@"保存成功"];
    }else {
        [SVProgressHUD showErrorWithStatus:@"保存失败"];
    }
}

#pragma mark - 构造函数
+ (instancetype)photoBroswerWithSelectedIndex:(NSInteger)selectedIndex urls:(NSArray<NSURL *> *)urls parentImageViews:(NSArray<UIImageView *> *)parentImageViews {
    return [[self alloc] initWithSelectedIndex:selectedIndex urls:urls parentImageViews:parentImageViews];
}

- (instancetype)initWithSelectedIndex:(NSInteger)selectedIndex urls:(NSArray<NSURL *> *)urls parentImageViews:(NSArray<UIImageView *> *)parentImageViews {
    
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        _photos = [[PictureBrowserPhotos alloc] init];
        _photos.selectedIndex = selectedIndex;
        _photos.urls = urls;
        _photos.parentImageViews = parentImageViews;
        
        self.modalPresentationStyle = UIModalPresentationCustom;
        _animator = [[PictureBrowserAnimator alloc] initWithPhotos:_photos];
        self.transitioningDelegate = _animator;
    }
    
    return self;
}

#pragma mark - 视图生命周期
- (void)viewDidLoad {
    [super viewDidLoad];
    
    // 滚动到选中图片
    [_collectionView scrollToItemAtIndexPath:[NSIndexPath indexPathForItem:_photos.selectedIndex inSection:0] atScrollPosition:UICollectionViewScrollPositionCenteredHorizontally animated:NO];
}

- (void)loadView {
    CGRect rect = [UIScreen mainScreen].bounds;
    rect.size.width += 20;
    self.view = [[UIView alloc] initWithFrame:rect];
    self.view.backgroundColor = [UIColor blackColor];
    
    [self setupUI];
}

#pragma mark - 懒加载控件
- (UICollectionView *)collectionView {
    if (!_collectionView) {
        _collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:[[PictureBrowserViewLayout alloc] init]];
    }
    
    return _collectionView;
}

- (UIButton *)pageConutButton {
    if(!_pageConutButton) {
        _pageConutButton = [[UIButton alloc] init];
    }
    
    return _pageConutButton;
}

#pragma mark - UICollectionViewDataSource
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return _photos.urls.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    PictureBrowserViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:PictureBrowserViewCellId forIndexPath:indexPath];
    
    cell.imageUrl = _photos.urls[indexPath.row];
    cell.pictureDelegate = self;
    
    return cell;
}

#pragma mark - UICollectionViewDelegate
- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    NSInteger page = scrollView.contentOffset.x / self.view.bounds.size.width;
    
    [self setPageCountWithIndex:page + 1];
}

- (void)setPageCountWithIndex:(NSInteger)index {
    [_pageConutButton setTitle:[NSString stringWithFormat:@"%ld/%lu",(long)index,(unsigned long)_photos.urls.count] forState:UIControlStateNormal];
}

#pragma mark - PictureBrowserCellDelegate
- (void)pictureBrowserCellWillDismiss {
    [self close];
}

- (void)pictureDidScrollToScaleWihScale:(CGFloat)scale {
    BOOL isHidden = scale < 1;
    _pageConutButton.hidden = isHidden;
    self.view.backgroundColor = isHidden ? [UIColor clearColor] : [UIColor blackColor];
    
    if (scale < 1) {
        self.view.alpha = scale;
        self.view.transform = CGAffineTransformMakeScale(scale, scale);
    }else {
        self.view.alpha = 1;
        self.view.transform = CGAffineTransformIdentity;
    }
}

#pragma mark - 设置界面
- (void)setupUI {
    // 添加控件
    [self.view addSubview:self.collectionView];
    [self.view addSubview:self.pageConutButton];
    
    // 设置位置
    _collectionView.frame = self.view.bounds;
    
    CGPoint center = self.view.center;
    _pageConutButton.frame = CGRectMake(0, 0, 80, 40);
    center.y = _pageConutButton.frame.size.height;
    _pageConutButton.center = center;
    
    // 准备页数显示按钮
    [self preparePageCountButton];
    
    // 准备collectionview
    [self prepareCollectionView];
}

- (void)preparePageCountButton {
    _pageConutButton.titleLabel.font = [UIFont boldSystemFontOfSize:20];
    [_pageConutButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    _pageConutButton.layer.cornerRadius = 6;
    _pageConutButton.backgroundColor = [UIColor colorWithWhite:0.01 alpha:0.5];
    
    [self setPageCountWithIndex:_photos.selectedIndex + 1];
}

- (void)prepareCollectionView {
    _collectionView.backgroundColor = [UIColor clearColor];

    [_collectionView registerClass:[PictureBrowserViewCell class] forCellWithReuseIdentifier:PictureBrowserViewCellId];
    
    _collectionView.dataSource = self;
    _collectionView.delegate = self;
}

@end

@implementation PictureBrowserViewLayout

- (void)prepareLayout {
    [super prepareLayout];
    
    self.itemSize = self.collectionView.bounds.size;
    self.minimumLineSpacing = 0;
    self.minimumInteritemSpacing = 0;
    self.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    
    self.collectionView.showsHorizontalScrollIndicator = NO;
    self.collectionView.pagingEnabled = YES;
    self.collectionView.bounces = NO;
}

@end

File: 仿写新浪微博/Classes/Tools/PictureBrowser/PictureBrowserViewController.swift
//
//  PictureBrowserViewController.swift
//  仿写新浪微博
//
//  Created by Apple on 2017/8/24.
//  Copyright © 2017年 WRQ. All rights reserved.
//

import UIKit
import SVProgressHUD

/// 可重用cellId
let PictureBrowserViewCellId = "PictureBrowserViewCellId"

class PictureBrowserViewController: UIViewController {
    /// 选中图片索引
    var indexPath: IndexPath
    /// 选中图片数组
    var thumbnailUrls: [URL]
    
    // MARK: - 监听方法
    @objc fileprivate func close() {
        dismiss(animated: true, completion: nil)
    }
    
    @objc fileprivate func save() {
        // 获取图片
        let cell = collectionView.visibleCells[0] as! PictureBrowserViewCell
        guard let image = cell.imageView.image else {
            return
        }
        
        // 保存图片
        UIImageWriteToSavedPhotosAlbum(image, self, #selector(PictureBrowserViewController.image(image:didFinishSavingWithError:contextInfo:)), nil)
        
    }
    
    @objc private func image(image: UIImage, didFinishSavingWithError error: NSError?, contextInfo: Any?) {
        let message = (error == nil) ? "保存成功" : "保存失败"
        SVProgressHUD.showSuccess(withStatus: message)
    }

    // MARK: - 构造函数
    init(indexPath: IndexPath, thumbnailUrls: [URL]) {
        self.indexPath = indexPath
        self.thumbnailUrls = thumbnailUrls
        
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: - 控制器生命周期
    override func loadView() {
        var rect = UIScreen.main.bounds
        rect.size.width += 20
        view = UIView(frame: rect)
        
        setupUI()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        collectionView.scrollToItem(at: indexPath, at: .centeredHorizontally, animated: false);
    }
    
    // MARK: - 懒加载控件
    fileprivate lazy var collectionView: UICollectionView = UICollectionView(frame: CGRect.zero, collectionViewLayout: PictureBrowserViewLayout())
    fileprivate lazy var saveButton: UIButton = UIButton(title: "保存", imageName: nil, backColor: UIColor.darkGray, fontSize: 14, color: UIColor.white)
    fileprivate lazy var closeButton: UIButton = UIButton(title: "关闭", imageName: nil, backColor: UIColor.darkGray, fontSize: 14, color: UIColor.white)
    
    /// 自定义布局
    class PictureBrowserViewLayout: UICollectionViewFlowLayout {
        override func prepare() {
            super.prepare()
            
            itemSize = collectionView!.bounds.size
            minimumLineSpacing = 0
            minimumInteritemSpacing = 0
            scrollDirection = .horizontal
            
            collectionView?.showsHorizontalScrollIndicator = false
            collectionView?.isPagingEnabled = true
            collectionView?.bounces = false
        }
    }
}

// MARK: - UICollectionViewDataSource
extension PictureBrowserViewController: UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return thumbnailUrls.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: PictureBrowserViewCellId, for: indexPath) as! PictureBrowserViewCell
        
        cell.imageUrl = thumbnailUrls[indexPath.item]
        cell.pictureDelegate = self
        
        return cell
    }
}

// MARK: - PictureBrowserCellDelegate
extension PictureBrowserViewController: PictureBrowserCellDelegate {
    func pictureBrowserViewCellDidTapImage() {
        close()
    }
}

extension PictureBrowserViewController: PictureBrowserDismissDelegate {
    func imageViewForAnimation() -> UIImageView {
        let imageView = UIImageView()
        
        let cell = collectionView.visibleCells[0] as! PictureBrowserViewCell
        imageView.image = cell.imageView.image
        
        imageView.frame = cell.scrollView.convert(cell.imageView.frame, to: UIApplication.shared.keyWindow)
        
//        // 测试
//        UIApplication.shared.keyWindow?.addSubview(imageView)
        
        return imageView
    }
    
    func indexPathForAnimation() -> IndexPath {
        return collectionView.indexPathsForVisibleItems[0]
    }
}

// MARK: - 设置界面
fileprivate extension PictureBrowserViewController {
    func setupUI() {
        // 添加控件
        view.addSubview(collectionView)
        view.addSubview(saveButton)
        view.addSubview(closeButton)
        
        // 设置布局
        let margin = UIScreen.main.scale * 4
        
        collectionView.frame = view.bounds
        
        closeButton.snp.makeConstraints { (make) in
            make.bottom.equalTo(view.snp.bottom).offset(-margin)
            make.left.equalTo(view.snp.left).offset(margin)
            make.size.equalTo(CGSize(width: 100, height: 36))
        }
        
        saveButton.snp.makeConstraints { (make) in
            make.bottom.equalTo(view.snp.bottom).offset(-margin)
            make.right.equalTo(view.snp.right).offset(-margin)
            make.size.equalTo(CGSize(width: 100, height: 36))
        }
        
        // 监听方法
        closeButton.addTarget(self, action: #selector(PictureBrowserViewController.close), for: .touchUpInside)
        saveButton.addTarget(self, action: #selector(PictureBrowserViewController.save), for: .touchUpInside)
        
        // 准备collectionview
        prepareCollectionView()
    }
    
    func prepareCollectionView() {
        collectionView.register(PictureBrowserViewCell.self, forCellWithReuseIdentifier: PictureBrowserViewCellId)
        
        collectionView.dataSource = self
    }
}
